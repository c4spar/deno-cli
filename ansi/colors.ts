import * as stdColors from "https://deno.land/std@0.82.0/fmt/colors.ts";

type ExcludedColorMethods = "setColorEnabled" | "getColorEnabled";
type PropertyNames = keyof typeof stdColors;
type ColorMethod = (str: string, ...args: Array<unknown>) => string;
type ColorMethods = Exclude<PropertyNames, ExcludedColorMethods>;
type Chainable<T, E extends keyof T | null = null> = {
  [P in keyof T]: P extends E ? T[P] : Chainable<T, E> & T[P];
};

/** Chainable colors instance returned by all ansi escape properties. */
export type ColorsChain = Chainable<typeof stdColors, ExcludedColorMethods>;

/** Create new `Colors` instance. */
export type ColorsFactory = () => Colors;

/**
 * Chainable colors module.
 * If invoked as method, a new `Colors` instance will be returned.
 */
export type Colors = ColorsFactory & ColorsChain;

/**
 * Chainable colors module.
 * ```
 * console.log(colors.blue.bgRed.bold('Welcome to Deno.Land!'));
 * ```
 * If invoked as method, a new Ansi instance will be returned.
 * ```
 * const myColors: Colors = colors();
 * console.log(myColors.blue.bgRed.bold('Welcome to Deno.Land!'));
 * ```
 */
export const colors: Colors = factory();

function factory(): Colors {
  let stack: Array<ColorMethods> = [];

  const colors: Colors = function (
    this: ColorsChain | undefined,
    str: string,
    ...args: Array<unknown>
  ): string | ColorsChain {
    if (this) {
      const lastname = stack.pop() as ColorMethods;
      const method = stdColors[lastname] as ColorMethod;
      const result: string = method(
        stack.reduce(
          (str: string, name) => (stdColors[name] as ColorMethod)(str),
          str,
        ),
        ...args,
      );
      stack = [];
      return result;
    }
    return factory();
  } as Colors;

  const methodNames = Object.keys(stdColors) as Array<PropertyNames>;

  for (const name of methodNames) {
    if (name === "setColorEnabled" || name === "getColorEnabled") {
      continue;
    }
    Object.defineProperty(colors, name, {
      get(this: ColorsChain) {
        stack.push(name);
        return this;
      },
    });
  }

  return colors;
}
