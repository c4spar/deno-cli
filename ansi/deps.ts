export { encodeBase64 } from "https://deno.land/std@0.216.0/encoding/base64.ts";
export type {
  ReaderSync,
  WriterSync,
} from "https://deno.land/std@0.216.0/io/types.ts";
