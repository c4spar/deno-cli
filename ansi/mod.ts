export * from "./ansi.ts";
export * from "./ansi_escapes.ts";
export * from "./tty.ts";
