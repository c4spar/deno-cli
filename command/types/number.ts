import { number } from "../../flags/types/number.ts";
import { Type } from "../type.ts";
import { ITypeInfo } from "../types.ts";

export class NumberType extends Type<number> {
  public parse(type: ITypeInfo): number {
    return number(type);
  }
}
