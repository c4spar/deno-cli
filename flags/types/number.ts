import type { ITypeInfo, ITypeHandler } from "../types.ts";

export const number: ITypeHandler<number> = (
  { label, name, value, type }: ITypeInfo,
): number => {
  if (isNaN(Number(value))) {
    throw new Error(
      `${label} ${name} must be of type ${type} but got: ${value}`,
    );
  }

  return parseFloat(value);
};
