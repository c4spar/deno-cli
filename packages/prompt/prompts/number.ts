import { KeyEvent } from '../../keycode/lib/key-event.ts';
import { GenericInput, GenericInputPromptOptions } from './generic-input.ts';

export interface NumberPromptOptions extends GenericInputPromptOptions<number> {
    min?: number;
    max?: number;
    float?: boolean;
    round?: number;
}

export interface NumberPromptSettings extends NumberPromptOptions {
    min: number;
    max: number;
    float: boolean;
    round: number;
}

export class Number extends GenericInput<number, NumberPromptOptions, NumberPromptSettings> {

    public static async prompt( options: NumberPromptOptions ): Promise<number | undefined> {

        return new this( {
            min: -Infinity,
            max: Infinity,
            float: false,
            round: 2,
            ...options
        } ).run();
    }

    protected async handleEvent( event: KeyEvent ): Promise<boolean> {

        switch ( event.name ) {

            case 'c':
                if ( event.ctrl ) {
                    return Deno.exit( 0 );
                }
                break;

            case 'up':
                this.increaseValue();
                break;

            case 'down':
                this.decreaseValue();
                break;

            case 'left':
                this.moveCursorLeft();
                break;

            case 'right':
                this.moveCursorRight();
                break;

            case 'delete':
                this.deleteChar();
                break;

            case 'backspace':
                this.removeChar();
                break;

            case 'return':
            case 'enter':
                return this.selectValue();

            default:
                if ( event.sequence ) {
                    this.addChar( event.sequence );
                }
                break;
        }

        return false;
    }

    protected manipulateIndex( decrease?: boolean ) {

        if ( this.input[ this.index ] === '-' ) {
            this.index++;
        }

        if ( this.input.length && ( this.index > this.input.length - 1 ) ) {
            this.index--;
        }

        const decimalIndex: number = this.input.indexOf( '.' );
        const [ abs, dec ] = this.input.split( '.' );

        if ( dec && this.index === decimalIndex ) {
            this.index--;
        }

        const inDecimal: boolean = decimalIndex !== -1 && this.index > decimalIndex;
        let value: string = ( inDecimal ? dec : abs ) || '0';
        const oldLength: number = this.input.length;
        const index: number = inDecimal ? this.index - decimalIndex - 1 : this.index;
        const increaseValue = Math.pow( 10, value.length - index - 1 );

        value = ( parseInt( value ) + ( decrease ? -increaseValue : increaseValue ) ).toString();

        this.input = !dec ? value : ( this.index > decimalIndex ? abs + '.' + value : value + '.' + dec );

        if ( this.input.length > oldLength ) {
            this.index++;
        } else if ( this.input.length < oldLength && this.input[ this.index - 1 ] !== '-' ) {
            this.index--;
        }

        this.index = Math.max( 0, Math.min( this.index, this.input.length - 1 ) );
    }

    public increaseValue() {
        this.manipulateIndex( false );
    }

    public decreaseValue() {
        this.manipulateIndex( true );
    }

    protected addChar( char: string ): void {

        if ( !isNaN( char as any ) ) {
            super.addChar( char );
        } else if (
            this.settings.float &&
            char === '.' &&
            this.input.indexOf( '.' ) === -1 &&
            ( this.input[ 0 ] === '-' ? this.index > 1 : this.index > 0 )
        ) {
            super.addChar( char );
        }
    }

    protected sanitize( value: string ): number | undefined {

        if ( !value || isNaN( value as any ) ) {
            return;
        }

        const val: number = parseFloat( value );

        if ( this.settings.float ) {
            return parseFloat( val.toFixed( this.settings.round ) );
        }

        return val;
    }

    protected validate( value: number | undefined ): boolean | string {

        if ( typeof value !== 'number' ) {
            return false;
        }

        if ( value > this.settings.max ) {
            return `Value must be lower or equal than ${ this.settings.max }`;
        }

        if ( value < this.settings.min ) {
            return `Value must be greater or equal than ${ this.settings.min }`;
        }

        return true;
    }

    protected transform( value: number ): number {
        return value;
    }

    // protected getChar( index: number = this.index ): string {
    //     return this.input.slice( index, index + 1 );
    // }
    //
    // protected setChar( value: string | number, index: number = this.index ) {
    //     this.input = this.input.slice( 0, index ) + value.toString() + this.input.slice( index + 1 );
    // }
    //
    // public parseValue() {
    //     if ( this.input.length === 0 ) {
    //         return 0;
    //     }
    //
    //     const value: number = parseFloat( this.input );
    //
    //     if ( value < 0 ) {
    //
    //     }
    //
    //     const length = this.input.length;
    //     const value = parseFloat( this.getChar() ) || 0;
    //     this.input = this.input.slice( 0, this.index ) + ( value - 1 ) + this.input.slice( this.index + 1 );
    // }
}
