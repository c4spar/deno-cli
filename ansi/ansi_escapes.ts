import { encodeBase64 } from "./deps.ts";

/** Image options. */
export interface ImageOptions {
  width?: number;
  height?: number;
  preserveAspectRatio?: boolean;
}

/** Escape sequence: `\x1B` */
const ESC = "\x1B";
/** Control sequence intro: `\x1B[` */
const CSI = `${ESC}[`;
/** Operating system command: `\x1B]` */
const OSC = `${ESC}]`;
/** Link separator */
const SEP = ";";

export const ansiEscapes = {
  /** Ring audio bell: `\u0007` */
  bel: "\u0007",
  /** Get cursor position. */
  cursorPosition: `${CSI}6n`,
  /**
   * Move cursor to x, y, counting from the top left corner.
   * @param x Position left.
   * @param y Position top.
   */
  cursorTo(x: number, y?: number): string {
    if (typeof y !== "number") {
      return `${CSI}${x}G`;
    }
    return `${CSI}${y};${x}H`;
  },
  /**
   * Move cursor by offset.
   * @param x Offset left.
   * @param y Offset top.
   */
  cursorMove(x: number, y: number): string {
    let ret = "";

    if (x < 0) {
      ret += `${CSI}${-x}D`;
    } else if (x > 0) {
      ret += `${CSI}${x}C`;
    }

    if (y < 0) {
      ret += `${CSI}${-y}A`;
    } else if (y > 0) {
      ret += `${CSI}${y}B`;
    }

    return ret;
  },
  /**
   * Move cursor up by n lines.
   * @param count Number of lines.
   */
  cursorUp: (count = 1): string => `${CSI}${count}A`,
  /**
   * Move cursor down by n lines.
   * @param count Number of lines.
   */
  cursorDown: (count = 1): string => `${CSI}${count}B`,
  /**
   * Move cursor forward by n lines.
   * @param count Number of lines.
   */
  cursorForward: (count = 1): string => `${CSI}${count}C`,
  /**
   * Move cursor backward by n lines.
   * @param count Number of lines.
   */
  cursorBackward: (count = 1): string => `${CSI}${count}D`,
  /**
   * Move cursor to the beginning of the line n lines down.
   * @param count Number of lines.
   */
  cursorNextLine: (count = 1): string => `${CSI}E`.repeat(count),
  /**
   * Move cursor to the beginning of the line n lines up.
   * @param count Number of lines.
   */
  cursorPrevLine: (count = 1): string => `${CSI}F`.repeat(count),
  /** Move cursor to first column of current row. */
  cursorLeft: `${CSI}G`,
  /** Hide cursor. */
  cursorHide: `${CSI}?25l`,
  /** Show cursor. */
  cursorShow: `${CSI}?25h`,
  /** Save cursor. */
  cursorSave: `${ESC}7`,
  /** Restore cursor. */
  cursorRestore: `${ESC}8`,
  /**
   * Scroll window up by n lines.
   * @param count Number of lines.
   */
  scrollUp: (count = 1): string => `${CSI}S`.repeat(count),
  /**
   * Scroll window down by n lines.
   * @param count Number of lines.
   */
  scrollDown: (count = 1): string => `${CSI}T`.repeat(count),
  /** Clear screen. */
  eraseScreen: `${CSI}2J`,
  /**
   * Clear screen up.
   * @param count Number of lines.
   */
  eraseUp: (count = 1): string => `${CSI}1J`.repeat(count),
  /**
   * Clear screen down.
   * @param count Number of lines.
   */
  eraseDown: (count = 1): string => `${CSI}0J`.repeat(count),
  /** Clear current line. */
  eraseLine: `${CSI}2K`,
  /** Clear to line end. */
  eraseLineEnd: `${CSI}0K`,
  /** Clear to line start. */
  eraseLineStart: `${CSI}1K`,
  /**
   * Clear screen and move cursor by n lines up and move cursor to first column.
   * @param count Number of lines.
   */
  eraseLines(count: number): string {
    let clear = "";
    for (let i = 0; i < count; i++) {
      clear += ansiEscapes.eraseLine +
        (i < count - 1 ? ansiEscapes.cursorUp() : "");
    }
    clear += ansiEscapes.cursorLeft;
    return clear;
  },
  /**
   * Create link.
   * @param text Link text.
   * @param url Link url.
   * ```
   * console.log(
   *   ansi.link("Click me.", "https://deno.land"),
   * );
   * ```
   */
  link: (text: string, url: string): string =>
    [
      OSC,
      "8",
      SEP,
      SEP,
      url,
      ansiEscapes.bel,
      text,
      OSC,
      "8",
      SEP,
      SEP,
      ansiEscapes.bel,
    ].join(""),
  /**
   * Create image.
   * @param buffer  Image buffer.
   * @param options Image options.
   * ```
   * const response = await fetch("https://deno.land/images/hashrock_simple.png");
   * const imageBuffer: ArrayBuffer = await response.arrayBuffer();
   * console.log(
   *   ansi.image(imageBuffer),
   * );
   * ```
   */
  image(
    buffer: string | ArrayBuffer,
    options?: ImageOptions,
  ): string {
    let ret = `${OSC}1337;File=inline=1`;

    if (options?.width) {
      ret += `;width=${options.width}`;
    }

    if (options?.height) {
      ret += `;height=${options.height}`;
    }

    if (options?.preserveAspectRatio === false) {
      ret += ";preserveAspectRatio=0";
    }

    return ret + ":" + encodeBase64(buffer) + ansiEscapes.bel;
  },
};
