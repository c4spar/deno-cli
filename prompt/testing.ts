import { AssertionError } from "https://deno.land/std@0.170.0/testing/asserts.ts";
import { assertSnapshot } from "https://deno.land/std@0.170.0/testing/snapshot.ts";
import { eraseDown } from "../ansi/ansi_escapes.ts";
import { basename } from "./deps.ts";

export interface AssertPromptSnapshotOptions {
  /** Test name. */
  name: string;
  /** Import meta. Required to determine the import url of the test file. */
  meta: ImportMeta;
  /**
   * Object of test steps. Key is the test name and the value is an array of
   * input sequences/characters.
   */
  steps: Record<string, Array<string>>;
  /** Test function. */
  fn(): Promise<unknown>;
  /**
   * Operating system snapshot suffix. This is useful when your test produces
   * different output on different operating systems.
   */
  osSuffix?: Array<typeof Deno.build.os>;
  /**
   * Arguments passed to the `deno test` command when executing the snapshot
   * tests. `--allow-env=__CLIFFY_TEST_NAME__` is passed by default.
   */
  args?: Array<string>;
  /**
   * Enable/disable colors. Default is `false`.
   */
  colors?: boolean;
}

const encoder = new TextEncoder();

/**
 * Run prompt snapshot tests.
 *
 * ```ts
 * import { assertPromptSnapshot } from "./testing.ts";
 * import { Input } from "./input.ts";
 *
 * await assertPromptSnapshot({
 *   name: "test name",
 *   meta: import.meta,
 *   osSuffix: ["windows"],
 *   args: [],
 *   steps: {
 *     "should enter som text": ["foo bar", "\n"],
 *   },
 *   async fn() {
 *     await Input.prompt({
 *       message: "Whats your name?",
 *       default: "foo",
 *     });
 *   },
 * });
 * ```
 *
 * @param options Test options
 */
export async function assertPromptSnapshot(
  options: AssertPromptSnapshotOptions,
): Promise<void> {
  if (options.meta.main) {
    const testName = Deno.env.get("__CLIFFY_TEST_NAME__");
    if (testName === options.name) {
      await options.fn();
    }
  } else {
    registerTest(options);
  }
}

function registerTest(options: AssertPromptSnapshotOptions) {
  const fileName = basename(options.meta.url);

  Deno.test({
    name: options.name,
    async fn(ctx) {
      const tests = Object.entries(options.steps ?? {});
      if (!tests.length) {
        throw new Error(`No tests defined for: ${options.meta.url}`);
      }

      for (const [name, inputs] of tests) {
        await ctx.step({
          name,
          async fn(stepCtx) {
            const output: string = await runPrompt(inputs, options);
            const suffix = options.osSuffix?.includes(Deno.build.os)
              ? `.${Deno.build.os}`
              : "";

            await assertSnapshot(stepCtx, output, {
              path: `__snapshots__/${fileName}${suffix}.snap`,
            });
          },
        });
      }
    },
  });
}

async function runPrompt(
  inputs: Array<string>,
  options: AssertPromptSnapshotOptions,
): Promise<string> {
  try {
    const cmd = new Deno.Command("deno", {
      stdin: "piped",
      stdout: "piped",
      stderr: "piped",
      args: [
        "run",
        "--allow-env=__CLIFFY_TEST_NAME__",
        ...options.args ?? [],
        options.meta.url,
      ],
      env: {
        __CLIFFY_TEST_NAME__: options.name ?? "",
        NO_COLOR: options.colors ? "false" : "true",
      },
    });
    const child: Deno.ChildProcess = cmd.spawn();
    const writer = child.stdin.getWriter();

    for (const input of inputs) {
      await writer.write(encoder.encode(input));
      // Ensure all inputs are processed and rendered separately.
      await new Promise((resolve) => setTimeout(resolve, 600));
    }

    const { success, stdout, stderr, code } = await child.output();
    writer.releaseLock();
    await child.stdin.close();

    if (!success) {
      throw new Error(
        `Test command failed with a none zero exit code: ${code}. ${stderr}`,
      );
    }

    // Add a line break after each test input.
    return "stdout:\n" + new TextDecoder().decode(stdout).replaceAll(
      eraseDown(),
      eraseDown() + "\n",
    ) + "\nstderr:\n" + new TextDecoder().decode(stderr).replaceAll(
      eraseDown(),
      eraseDown() + "\n",
    );
  } catch (error: unknown) {
    const assertionError = new AssertionError(
      `Test failed: ${options.meta.url}`,
    );
    assertionError.cause = error;
    throw error;
  }
}
