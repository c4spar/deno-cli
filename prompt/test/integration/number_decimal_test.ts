import { ansi } from "../../../ansi/ansi.ts";
import { Number } from "../../number.ts";
import { assertPromptSnapshot } from "../../testing.ts";

await assertPromptSnapshot({
  name: "number prompt with float",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should enter a floating number": ansi
      .text("19.")
      .cursorUp
      .cursorUp
      .cursorUp
      .cursorUp
      .text("\n")
      .toArray(),
  },
  async fn() {
    await Number.prompt({
      message: "How old are you?",
      default: 7,
      max: 20,
      min: 18,
      float: true,
    });
  },
});
