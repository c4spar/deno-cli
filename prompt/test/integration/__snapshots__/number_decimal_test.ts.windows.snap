export const snapshot = {};

snapshot[`number prompt with float > should enter a floating number 1`] = `
"stdout:
? How old are you? (7) » \\x1b[26G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.1\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.2\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.3\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.4\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.4
\\x1b[?25h\\x1b[?25h
stderr:
"
`;
